### ISSUE TITLE
Can't export unused PackageReference to generated .nuspec

### ISSUE THREAD

**Issue by mythz** [NuGet/Home#7631]
I'm trying to migrate to using MSBuild Pack support for using .csproj to building NuGet package where I'm using local .dll's to build the project which I want to replace/swap to reference an external NuGet package in the generated .nuspec. 

The closest documented example of this I could find is [Replacing one library from a restore graph
](https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#replacing-one-library-from-a-restore-graph) where it suggests you can replace an external NuGet reference:

```xml
<PackageReference Include="Newtonsoft.Json" Version="9.0.1">
  <ExcludeAssets>All</ExcludeAssets>
</PackageReference>
```

With a reference to a local .dll:

```xml
<Reference Include="Newtonsoft.Json.dll" />
```

I'm trying to do something similar where I want to build against my local `.dll's` that were injected as an artifact from a dependent TeamCity/CI build:

```xml
<Reference Include="..\..\lib\net45\ServiceStack.Interfaces.dll" />
<Reference Include="..\..\lib\net45\ServiceStack.Text.dll" />
<Reference Include="..\..\lib\net45\ServiceStack.Common.dll" />
```

But when using `ExcludeAssets=All` as per the documentation:

```xml
<PackageReference Include="ServiceStack.Common" Version="5.0.0">
  <ExcludeAssets>All</ExcludeAssets>
</PackageReference>
```

It doesn't get exported in the package dependency list, e.g:

```xml
<group targetFramework=".NETFramework4.5" />
```

The closest I've come to getting the preferred behavior is to use `ExcludeAssets="compile"` so my local project doesn't build against it, except this also gets exported in the .nuspec:

```xml
<group targetFramework=".NETFramework4.5">
  <dependency id="ServiceStack.Common" version="5.4.0" exclude="Compile,Build,Analyzers" />
</group>
```

I only want to avoid building against it locally but have it exported as a normal dependency. The other issue with this approach is that I need it to reference a package that hasn't been published to NuGet yet (as all packages are published in lock-step), e.g:

```xml
<PackageReference Include="ServiceStack.Common" Version="5.5.0" ExcludeAssets="compile"/>
```

The build fails that it can't find the unpublished package:

    [NU1102] Unable to find package ServiceStack.Common with version (>= 5.5.0)

I effectively need someway to "inject" the exact dependencies and version I need in the generated .nuspec which results in:

```xml
<group targetFramework="net45">
  <dependency id="ServiceStack.Common" version="5.5.0" />
</group>
<group targetFramework=".netstandard2.0">
  <dependency id="ServiceStack.Common" version="5.5.0" />
</group>
```

Is there some way I can manually declare `<dependency/>` as above so it's only used when MSBuild/NuGet packs the project? Otherwise is there a way to apply some XML transform to the generated .nuspec so I can manipulate the XML in the .nuspec before it's packed/compressed?

Basically I'm only interested in injecting dependencies when the "pack" target is run so it's ignored/inert in all other targets.

**Comment by jainaashish** [comment]
You can have an explicit `.nuspec` file with your desired behavior and use that to pack your project. But besides that, I don't see this is a valid usecase in itself, what you're saying is PackageB wont consume assembly from PackageC but any one depending on PackageB should consume PackageC assembly. So I'm not sure why you have such requirements.

If your requirement is that you have our own `PackageC.dll` and wants to make it part of PackageB nuspec then you can simply pack that explicit assembly reference. You don't need a PackageReference to PackageC in PackageB for that.

**Comment by mythz** [comment]
> You can have an explicit .nuspec file with your desired behavior and use that to pack your project.

I'm already currently using a custom .nuspec file but the goal is to migrate to use MSBuild generated .nupkg. When I tried using a custom `NuspecFile` it overrode the entire generated .nuspec, I just want to be able to inject the missing dependencies into the MSBuild generated .nuspec. The catalyst to migrating to MSBuild pack is to use SourceLink which injects custom properties in the generated .nuspec.

> I don't see this is a valid usecase in itself, what you're saying is PackageB wont consume assembly from PackageC but any one depending on PackageB should consume PackageC assembly. 

I'm not sure what you're inferring here, [ServiceStack.Common](https://www.nuget.org/packages/ServiceStack.Common) has a dependency on ServiceStack.Text and ServiceStack.Interfaces so I don't have to declare individual package references for each, but I don't mind if I need to. 

Is there something I can add to the .dll reference to tell it what Package `<dependency/>` and Version it should generate in the .nuspec?

```xml
<Reference Include="..\..\lib\net45\ServiceStack.Common.dll" />
```

**Comment by jainaashish** [comment]
> I'm already currently using a custom .nuspec file but the goal is to migrate to use MSBuild generated .nupkg. When I tried using a custom NuspecFile it overrode the entire generated .nuspec, I just want to be able to inject the missing dependencies into the MSBuild generated .nuspec. The catalyst to migrating to MSBuild pack is to use SourceLink which injects custom properties in the generated .nuspec.

Even MSBuild or dotnet pack can also use an explicit nuspec file to generate nupkg.

> Is there something I can add to the .dll reference to tell it what Package <dependency/> and Version it should generate in the .nuspec?
Why would you generate a package dependency for an assembly reference? Just set `Pack=true` on that assembly reference which will be added in the generated nuspec as Assembly reference only. And will be added into the generated nupkg file. You can read more about assembly references in nuspec [here](https://docs.microsoft.com/en-us/nuget/reference/nuspec#framework-assembly-references)

**Comment by mythz** [comment]
>Why would you generate a package dependency for an assembly reference? Just set Pack=true on that assembly reference which will be added in the generated nuspec as Assembly reference only. And will be added into the generated nupkg file. You can read more about assembly references in nuspec 
> [here](https://docs.microsoft.com/en-us/nuget/reference/nuspec#framework-assembly-references)

The external .dll's allows us to maintain self-encapsulated projects where we can build external solutions and publish any changes in the local `/lib` folder. The .dll's used for development shouldn't be included in the package, they need to be replaced with the external NuGet package reference that they'll be published under.  The linked docs doesn't show anywhere that I can tell it what NuGet package `<dependency/>` it should use?

**Comment by StingyJack** [comment]
Linking to what I think is the same issue so as to increase the "vote" count -  https://github.com/NuGet/Home/issues/7371
